runtime/compiler/ilgen/Walker.cpp

```diff
@@ -7478,14 +7479,14 @@ TR_J9ByteCodeIlGenerator::storeArrayElement(TR::DataType dataType, TR::ILOpCodes
        !TR::Compiler->om.canGenerateArraylets() &&
        dataType == TR::Address)
       {
-      if (!TR::Compiler->om.isValueTypeArrayFlatteningEnabled() &&
-          _methodSymbol->skipNonNullableArrayNullStoreCheck())
+      if (TR::Compiler->om.isValueTypeArrayFlatteningEnabled())
          {
-         generateNonHelper = false;
+         generateNonHelper = (_methodSymbol->skipNonNullableArrayNullStoreCheck() && _methodSymbol->skipFlattenableArrayElementNonHelperCall())
+                                ? false : true;
          }
       else
          {
-         generateNonHelper = true;
+         generateNonHelper = _methodSymbol->skipNonNullableArrayNullStoreCheck() ? false : true;
```

Comment:
I find it a little difficult to follow this mix of ifs, short-circuiting operators, and ternary operators, which are all together performing a logical computation.  I wonder if it would be easier to read as

```
generateNonHelper = (TR::Compiler->om.isValueTypeArrayFlatteningEnabled()
                     && !_methodSymbol->skipFlattenableArrayElementNonHelperCall())
                    || !_methodSymbol->skipNonNullableArrayNullStoreCheck();
```

That is, "if flattening is enabled, and we're not able to skip using the non-helper call in this method, generate the non-helper, or if we can't avoid performing a check for a null store to a non-nullable array, generate the non-helper."